metagraph/src/common/algorithms.hpp

```diff
@@ -66,37 +66,137 @@ namespace utils {
+    // Return true if the two sorted ranges share a common element
+    template <class InIt1, class InIt2>
+    constexpr bool share_element(InIt1 a_begin, InIt1 a_end, InIt2 b_begin, InIt2 b_end) {
+        assert(std::is_sorted(a_begin, a_end));
+        assert(std::is_sorted(b_begin, b_end));
+
+        while (a_begin != a_end && b_begin != b_end) {
+            if (*a_begin < *b_begin) {
+                ++a_begin;
+            } else if (*b_begin < *a_begin) {
+                ++b_begin;
+            } else {
+                return true;
+            }
+        }
 
-            if (second_begin == second_end)
-                break;
+        return false;
+    }
 
-            if (*first_begin == *second_begin) {
-                ++count;
-                ++first_begin;
-                ++second_begin;
+    // Intersect the sorted ranges a1 and b1 with corresponding sorted ranges of
+    // sorted ranges a2 and b2 (of equal length).
+    // i.e., For each shared element between a1 and b1, intersect the corresponding
+    // ranges in a2 and b2.
+    template <class OutType, class SetOp,
+              class InIt1, class InIt2, class InIt3, class InIt4,
+              class OutIt1, class OutIt2, typename... Args>
+    constexpr std::tuple<size_t, size_t, size_t> indexed_set_op(InIt1 a1_begin,
+                                                                InIt1 a1_end,
+                                                                InIt2 a2_begin,
+                                                                InIt3 b1_begin,
+                                                                InIt3 b1_end,
+                                                                InIt4 b2_begin,
+                                                                OutIt1 out1,
+                                                                OutIt2 out2,
+                                                                Args&&... args) {
+        size_t a_size = 0;
+        size_t b_size = 0;
+        size_t new_size = 0;
+        SetOp set_op(std::forward<Args>(args)...);
+
+        while (a1_begin != a1_end && b1_begin != b1_end) {
+            if (*a1_begin < *b1_begin) {
+                ++a1_begin;
+                ++a2_begin;
+            } else if (*b1_begin < *a1_begin) {
+                ++b1_begin;
+                ++b2_begin;
+            } else {
+                a_size += a2_begin->size();
+                b_size += b2_begin->size();
+                OutType merged;
+                set_op(a2_begin->begin(), a2_begin->end(),
+                       b2_begin->begin(), b2_begin->end(),
+                       std::back_inserter(merged));
+                if (merged.size()) {
+                    new_size += merged.size();
+                    *out1 = *a1_begin;
+                    ++out1;
+                    *out2 = std::move(merged);
+                    ++out2;
+                }
```

Comment:
Hm, I don't quite get what's going on. If the ranges are disjoint, you skip it in the output completely and you never know which one was skipped.

Reply:
Hm, no, I'm totally confused by the description.